components/hab/src/scaffolding.rs

```diff
@@ -0,0 +1,167 @@
+                    Ok(Some(ident))
+                }
+                _ => {
+                    let ident = PackageIdent::from_str(scaffold).unwrap();
+                    let _ = ui.status(Status::Using, &format!("custom scaffolding: '{}'", ident));
+                    ui.para("")?;
+                    Ok(Some(ident))
+                }
+            }
+        }
+        // If nothing was passed try to autodiscover the codebase
+        // I'm not saying it's magic. But, MAGIC.
+        None => magic_function(ui),
+    }
+}
+fn magic_function(ui: &mut UI) -> Result<Option<PackageIdent>> {
+    // Determine if the current dir has an app that can use
+    // one of our scaffoldings and use it by default
+    ui.begin("Attempting autodiscovery ")?;
+    ui.para(
+        "No scaffolding type was provided. Let's see if we can figure out \
+        what kind of application you're planning to package.",
+    )?;
+    let current_path = Path::new(".");
+    if is_project_ruby(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-ruby").unwrap();
+        ui.begin("We've detected your app as Ruby")?;
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        ui.para("")?;
+        Ok(Some(ident))
+    } else if is_project_golang(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-go").unwrap();
+        ui.begin("We've detected your app as Golang")?;
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        ui.para("")?;
+        Ok(Some(ident))
+    } else if is_project_node(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-node").unwrap();
+        ui.begin("We've detected your app as Node.js")?;
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        ui.para("")?;
+        Ok(Some(ident))
+    } else {
+        ui.warn(
+            "Unable to determine the type of app in your current directory",
+        )?;
+        ui.para(
+            "For now, we'll generate a plan with all of the available plan options and callbacks. \
+            For more documentation on plan options try passing --withdocs or visit \
+            https://www.habitat.sh/docs/reference/plan-syntax/",
+        )?;
+        Ok(None)
+    }
+}
+
+fn is_project_ruby<T>(path: T) -> bool
+where
+    T: AsRef<Path>,
+{
+    if path.as_ref().join("Gemfile").is_file() {
+        return true;
+    }
+    return false;
+}
+
+fn is_project_golang<T>(path: T) -> bool
+where
+    T: AsRef<Path>,
+{
+    if path.as_ref().join("main.go").is_file() ||
+        path.as_ref().join("Godeps/Godeps.json").is_file() ||
+        path.as_ref().join("vendor/vendor.json").is_file() ||
+        path.as_ref().join("glide.yaml").is_file() ||
+        project_uses_gb(path.as_ref()).expect("Result<bool> not returned from .go file check")
+    {
+        return true;
+    }
+    return false;
```

Comment:
Like Ruby, the last line of a block in Rust returns that value. So this,

```
fn what() -> bool {
    true
}
```

Will return true. However,

```
fn what() -> bool {
    true;
}
```

This will result in a compilation error. Placing a `;` at the end of a statement will force it to return `()` ("unit").

The `return` isn't necessary to return a value on the last line, it's just used for early returns.

You don't need to make any changes here, I just wanted to give you this knowledge ;)